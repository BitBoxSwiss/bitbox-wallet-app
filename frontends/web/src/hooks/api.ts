/**
 * Copyright 2022 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DependencyList, useEffect, useState } from 'react';
import { TSubscriptionCallback, TUnsubscribe } from '../api/subscribe';
import { useMountedRef } from './mount';

/**
 * useSubscribeReset is a hook to subscribe to a subscription function.
 * starts on first render, and returns undefined while there is no first response.
 * re-renders on every update.
 * An array is returned: `[value, reset]`, where value is the subscribed value and `reset()` resets
 * the value to `undefined`.
 */
export const useSubscribeReset = <T>(
  subscription: ((callback: TSubscriptionCallback<T>) => TUnsubscribe),
): [T | undefined, () => void] => {
  const [response, setResponse] = useState<T>();
  const mounted = useMountedRef();
  const subscribe = () => {
    return subscription((data) => {
      if (mounted.current) {
        setResponse(data);
      }
    });
  };
  useEffect(
    () => subscribe(),
    // empty dependencies because it's only subscribed once
    [] // eslint-disable-line react-hooks/exhaustive-deps
  );
  return [response, () => setResponse(undefined)];
};

/**
 * useSubscribe is a hook to subscribe to a subscription function.
 * starts on first render, and returns undefined while there is no first response.
 * re-renders on every update.
 */
export const useSubscribe = <T>(
  subscription: ((callback: TSubscriptionCallback<T>) => TUnsubscribe),
): (T | undefined) => {
  const [response] = useSubscribeReset(subscription);
  return response;
};

/**
 * useLoad is a hook to load a promise.
 * gets fired on first render, and returns undefined while loading.
 * Optionally pass a dependency array as 2nd arguemnt to control re-executing apiCall
 */
export const useLoad = <T>(
  apiCall: (() => Promise<T>) | null,
  dependencies?: DependencyList,
): (T | undefined) => {
  const [response, setResponse] = useState<T>();
  const mounted = useMountedRef();
  const load = () => {
    if (apiCall === null) {
      return;
    }
    apiCall().then((data) => {
      if (mounted.current) {
        setResponse(data);
      }
    });
  };
  useEffect(
    () => load(),
    // By default no dependencies are passed to only query once
    dependencies || [] // eslint-disable-line react-hooks/exhaustive-deps
  );
  return response;
};

/**
 * useSync is a hook to load a promise and sync to a subscription function.
 * It is a combination of useLoad and useSubscribe.
 * gets fired on first render, and returns undefined while loading,
 * re-renders on every update.
 */
export const useSync = <T>(
  apiCall: () => Promise<T>,
  subscription: ((callback: TSubscriptionCallback<T>) => TUnsubscribe),
): (T | undefined) => {
  const [response, setResponse] = useState<T>();
  const mounted = useMountedRef();
  const onData = (data: T) => {
    if (mounted.current) {
      setResponse(data);
    }
  };
  useEffect(
    () => {
      apiCall().then(onData);
      return subscription(onData);
    }, // we pass no dependencies because it's only queried once
    []); // eslint-disable-line react-hooks/exhaustive-deps
  return response;
};
