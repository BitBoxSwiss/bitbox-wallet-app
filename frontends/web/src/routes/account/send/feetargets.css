.row {
    display: flex;
    flex-wrap: wrap;
}

.column {
    flex-basis: 50%;
    flex-grow: 1;
    flex-shrink: 1;
}
.column:first-child {
    flex-basis: 10%;
}

select.priority {
    border-right: none;
    padding-right: 0;
    text-transform: capitalize;
}

.fee input {
    border-left: none;
    padding-left: 0;
}

select:disabled.priority,
.fee input:disabled {
    border-color: var(--color-lightgray) !important;
}

select.priority,
.fee input {
    border-color: var(--color-mediumgray) !important;
}

.feeDescription {
    color: var(--color-secondary);
    flex: 0 0 50%;
    font-size: var(--size-small);
    font-weight: 400;
    margin-top: 0;
}
