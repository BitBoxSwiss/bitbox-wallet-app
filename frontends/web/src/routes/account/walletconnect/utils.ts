/**
 * Copyright 2023 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Core } from '@walletconnect/core';
import { IWeb3Wallet, Web3Wallet } from '@walletconnect/web3wallet';
import { useCallback, useEffect, useState } from 'react';

export let web3wallet: IWeb3Wallet;

export const SUPPORTED_CHAINS = [
  'eip155:1', // ETH Mainnet
  'eip155:5', // ETH Goerli testnet
  'eip155:10', // Optimism
  'eip155:56', // BSC
  'eip155:137', // Polygon
  'eip155:250', // Fantom
  'eip155:42161' // Arbitrum One
];

export const EIP155_SIGNING_METHODS = {
  PERSONAL_SIGN: 'personal_sign',
  ETH_SIGN: 'eth_sign',
  ETH_SIGN_TRANSACTION: 'eth_signTransaction',
  /**
   * Many dapps will request an 'eth_signTypedData' but in reality expect 'eth_signTypedData_v4'
   * because that's basically the standard. V4 is assumed, but not necessarily asked for specifically.
   * There aren't many uses of V3 in the wild, but leaving in just in case and returning a v4 signature instead,
   * the dapp can validate this anyway
   */
  ETH_SIGN_TYPED_DATA: 'eth_signTypedData',
  ETH_SIGN_TYPED_DATA_V3: 'eth_signTypedData_v3',
  ETH_SIGN_TYPED_DATA_V4: 'eth_signTypedData_v4',
  ETH_SEND_TRANSACTION: 'eth_sendTransaction'
};

// Initialize the Web3Wallet
export default function useInitialization() {
  const [initialized, setInitialized] = useState(false);
  const onInitialize = useCallback(async () => {
    try {
      await createWeb3Wallet();
      setInitialized(true);
    } catch (err: unknown) {
      console.log('Error for initializing', err);
    }
  }, []);

  useEffect(() => {
    if (!initialized) {
      onInitialize();
    }
  }, [onInitialize, initialized]);

  return initialized;
}


export async function createWeb3Wallet() {
  const core = new Core({
    projectId: '89733df088867a1a1bf644013addd6cc',
  });

  web3wallet = await Web3Wallet.init({
    core,
    metadata: {
      name: 'BitBoxApp',
      description: 'BitBoxApp',
      url: 'https://bitbox.swiss',
      // TODO: Add BitBox logo
      icons: []
    }
  });
}

function getTopicFromURI(wcURI: string) {
  try {
    // Split the URI by ':' and then by '@', and take the part before the '@'
    return wcURI.split(':')[1].split('@')[0];
  } catch {
    return '';
  }
}

// It is necessary to check the history for previously rejected pairing attempts.
// Without this check, a pairing URI that was previously rejected (and thus is invalid)
// might still be paired (for unknown reasons the WC library does this).

// This "invalid pairing" won't request a new session and
// also won't throw any error (on the first attempt) rendering it
// non functional and potentially confuses the user.
function pairingHasEverBeenRejected(topic: string) {
  return web3wallet.core.history.values.findIndex(history =>
    history.topic === topic &&
    history.response &&
    'error' in history.response)
    >= 0;
}

export async function pair(params: { uri: string }) {
  const { uri } = params;

  const topic = getTopicFromURI(uri);
  const hasEverBeenRejected = pairingHasEverBeenRejected(topic);
  if (hasEverBeenRejected) {
    throw new Error('Please use a new URI!');
  }
  await web3wallet?.core.pairing.pair({ uri });
}

export const getAddressFromEIPString = (address: string) => {
  const parts = address.split(':');
  return parts.length > 2 ? parts[2] : '';
};

export const truncateAddress = (address: string) => {
  if (!address) {
    return '';
  }
  return `${address.substring(0, 6)}...${address.substring(address.length - 6)}`;
};