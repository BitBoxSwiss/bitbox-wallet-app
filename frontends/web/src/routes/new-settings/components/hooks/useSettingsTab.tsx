import { Main, Header } from '../../../../components/layout';
import { View, ViewContent } from '../../../../components/view/view';
import Tabs from '../tabs';

const useSettingsTab = (Component: () => JSX.Element) => {
  const ComponentWithTabs = () => (
    <Main>
      <Header title={<h2>Settings</h2>} />
      <View fullscreen={false}>
        <ViewContent>
          <Tabs />
          <Component />
        </ViewContent>
      </View>
    </Main>
  );
  return ComponentWithTabs;
};


export default useSettingsTab;