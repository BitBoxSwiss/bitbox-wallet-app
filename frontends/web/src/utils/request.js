/**
 * Copyright 2018 Shift Devices AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import i18n from '../i18n/i18n';
import { extConfig } from './config';
import { alertUser } from '../components/alert/Alert';
import { call } from './qttransport';
import { androidCall } from './androidtransport';
import { runningInAndroid, runningInQtWebEngine } from './env';

export const apiPort = extConfig('{{ API_PORT }}', '8082');
export const apiToken = extConfig('{{ API_TOKEN }}', '');

export function isTLS() {
    return document.URL.startsWith('https://');
}

export function apiURL(endpoint) {
    return (isTLS() ? 'https://' : 'http://') + 'localhost:' + apiPort + '/api/' + endpoint;
}

function handleError(endpoint) {
    return function(json) {
        return new Promise((resolve, reject) => {
            if (json && json.error) {
                if (json.error.indexOf('hidapi: unknown failure') !== -1) {
                    // Ignore device communication errors. Usually
                    // happens when unplugged during an operation, in
                    // which case the result does not matter.
                    return;
                }
                console.error('error from endpoint', endpoint, json);
                alertUser(i18n.t('genericError'));
                reject(json.error);
                return;
            }
            resolve(json);
        });
    };
}

export function apiGet(endpoint) {
    if (runningInQtWebEngine()) {
        return call(JSON.stringify({
            method: 'GET',
            endpoint,
        }));
    }
    if (runningInAndroid()) {
        return androidCall(JSON.stringify({
            method: 'GET',
            endpoint,
        }));
    }
    return fetch(apiURL(endpoint), {
        method: 'GET'
    }).then(response => response.json()).then(handleError(endpoint));
}

export function apiPost(endpoint, body) {
    if (runningInQtWebEngine()) {
        return call(JSON.stringify({
            method: 'POST',
            endpoint,
            body: JSON.stringify(body)
        }));
    }
    if (runningInAndroid()) {
        return androidCall(JSON.stringify({
            method: 'POST',
            endpoint,
            body: JSON.stringify(body)
        }));
    }
    return fetch(apiURL(endpoint), {
        method: 'POST',
        body: JSON.stringify(body)
    }).then(response => response.json()).then(handleError(endpoint));
}
