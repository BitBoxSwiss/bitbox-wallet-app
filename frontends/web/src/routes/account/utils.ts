/**
 * Copyright 2018 Shift Devices AG
 * Copyright 2021 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CoinCode, ScriptType, IAccount, CoinUnit } from '../../api/account';

export function findAccount(accounts: IAccount[], accountCode: string): IAccount | undefined {
  return accounts.find(({ code }) => accountCode === code);
}

export function getCryptoName(cryptoLabel: string, account?: IAccount): string {
  if (account && isBitcoinOnly(account.coinCode)) {
    return 'Bitcoin';
  }
  return cryptoLabel;
}

export function isBitcoinOnly(coinCode: CoinCode): boolean {
  switch (coinCode) {
  case 'btc':
  case 'tbtc':
    return true;
  default:
    return false;
  }
}

export const isBitcoinCoin = (coin: CoinUnit) => (coin === 'BTC') || (coin === 'TBTC') || (coin === 'sat') || (coin === 'tsat');

export function isBitcoinBased(coinCode: CoinCode): boolean {
  switch (coinCode) {
  case 'btc':
  case 'tbtc':
  case 'ltc':
  case 'tltc':
    return true;
  default:
    return false;
  }
}

export function isEthereumBased(coinCode: CoinCode): boolean {
  return coinCode === 'eth' || coinCode === 'goeth' || coinCode === 'sepeth' || coinCode.startsWith('eth-erc20-');
}

export function getCoinCode(coinCode: CoinCode): CoinCode | undefined {
  switch (coinCode) {
  case 'btc':
  case 'tbtc':
    return 'btc';
  case 'ltc':
  case 'tltc':
    return 'ltc';
  case 'eth':
  case 'goeth':
  case 'sepeth':
    return 'eth';
  }
}


export function getScriptName(scriptType: ScriptType): string {
  switch (scriptType) {
  case 'p2pkh':
    return 'Legacy (P2PKH)';
  case 'p2wpkh-p2sh':
    return 'Wrapped Segwit (P2WPKH-P2SH)';
  case 'p2wpkh':
    return 'Native segwit (bech32, P2WPKH)';
  case 'p2tr':
    return 'Taproot (bech32m, P2TR)';
  }
}

export function customFeeUnit(coinCode: CoinCode): string {
  if (isBitcoinBased(coinCode)) {
    return 'sat/vB';
  }
  if (isEthereumBased(coinCode)) {
    return 'Gwei';
  }
  return '';
}
