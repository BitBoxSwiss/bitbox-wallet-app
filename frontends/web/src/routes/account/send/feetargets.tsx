/**
 * Copyright 2018 Shift Devices AG
 * Copyright 2021-2024 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState, useEffect, useRef, ChangeEvent, useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { useLoad } from '../../../hooks/api';
import * as accountApi from '../../../api/account';
import { getConfig } from '../../../utils/config';
import { Input, Select } from '../../../components/forms';
import { Message } from '../../../components/message/message';
import { customFeeUnit, getCoinCode, isEthereumBased } from '../utils';
import style from './feetargets.module.css';

type Props = {
    accountCode: accountApi.AccountCode;
    coinCode: accountApi.CoinCode;
    disabled: boolean;
    fiatUnit: accountApi.ConversionUnit;
    proposedFee?: accountApi.IAmount;
    customFee: string;
    showCalculatingFeeLabel?: boolean;
    onFeeTargetChange: (code: accountApi.FeeTargetCode) => void;
    onCustomFee: (customFee: string) => void;
    error?: string;
}


type TOptions = {
    value: accountApi.FeeTargetCode;
    text: string;
}

export const FeeTargets = ({
  accountCode,
  coinCode,
  disabled,
  fiatUnit,
  proposedFee,
  customFee,
  showCalculatingFeeLabel,
  onFeeTargetChange,
  onCustomFee,
  error
}: Props) => {
  const { t } = useTranslation();
  const config = useLoad(getConfig);
  const [feeTarget, setFeeTarget] = useState<string>('');
  const [options, setOptions] = useState<TOptions[] | null>(null);
  const [noFeeTargets, setNoFeeTargets] = useState<boolean>(false);

  const inputRef = useRef<HTMLInputElement & { autofocus: boolean }>(null);

  const focusInput = useCallback(() => {
    if (!disabled && inputRef.current && inputRef.current.autofocus) {
      inputRef.current.focus();
    }
  }, [disabled]);

  useEffect(() => {
    accountApi.getFeeTargetList(accountCode)
      .then(({ feeTargets, defaultFeeTarget }) => {
        if (config) {
          const expert = config.frontend.expertFee || feeTargets.length === 0;
          const options = feeTargets.map(({ code, feeRateInfo }) => ({
            value: code,
            text: t(`send.feeTarget.label.${code}`) + (expert && feeRateInfo ? ` (${feeRateInfo})` : ''),
          }));
          if (expert) {
            options.push({
              value: 'custom',
              text: t('send.feeTarget.label.custom'),
            });
          }
          setOptions(options);
          setFeeTarget(defaultFeeTarget);
          onFeeTargetChange(defaultFeeTarget);
          if (feeTargets.length === 0) {
            setNoFeeTargets(true);
          }
        }
      })
      .catch(console.error);
    focusInput();
  }, [t, focusInput, accountCode, config, onFeeTargetChange]);

  const handleFeeTargetChange = (event: React.SyntheticEvent) => {
    const target = event.target as HTMLSelectElement;
    const value = target.options[target.selectedIndex].value as accountApi.FeeTargetCode;
    setFeeTarget(value);
    onFeeTargetChange(value);
  };

  const handleCustomFee = (event: ChangeEvent<HTMLInputElement>) => {
    onCustomFee(event.target.value);
  };

  const getProposeFeeText = (): string => {
    if (!proposedFee) {
      return '';
    }
    const { amount, unit, conversions } = proposedFee;
    return `${amount} ${unit} ${conversions ? ` = ${conversions[fiatUnit === 'sat' ? 'BTC' : fiatUnit]} ${fiatUnit}` : ''}`;
  };

  if (options === null) {
    return (
      <Input
        label={t('send.priority')}
        id="feetarget"
        placeholder={t('send.feeTarget.placeholder')}
        disabled
        value=""
        transparent />
    );
  }

  const isCustom = feeTarget === 'custom';
  const hasOptions = options.length > 0;
  const proposeFeeText = getProposeFeeText();
  const preventFocus = document.activeElement && document.activeElement.nodeName === 'INPUT';

  return (
    hasOptions ? (
      <div>
        {!isCustom ? (
          showCalculatingFeeLabel ? (
            <Input
              disabled
              label={t('send.priority')}
              placeholder={t('send.feeTarget.placeholder')}
              value=""
              transparent />
          ) : (
            <Select
              className={style.priority}
              label={t('send.priority')}
              id="feeTarget"
              disabled={disabled}
              onChange={handleFeeTargetChange}
              value={feeTarget}
              options={options} />
          )
        ) : (
          <div className={style.rowCustomFee}>
            { noFeeTargets ? (
              <Message small type="warning">
                {t('send.noFeeTargets')}
              </Message>
            ) : null }
            <div className={style.column}>
              <Select
                className={style.priority}
                label={t('send.priority')}
                id="feeTarget"
                disabled={disabled}
                onChange={handleFeeTargetChange}
                value={feeTarget}
                options={options} />
            </div>
            <div className={style.column}>
              <Input
                type={disabled ? 'text' : 'number'}
                min="0"
                step="any"
                autoFocus={!preventFocus}
                align="right"
                className={`${style.fee} ${style.feeCustom}`}
                disabled={disabled}
                label={t('send.feeTarget.customLabel', {
                  context: isEthereumBased(coinCode) ? 'eth' : ''
                })}
                id="proposedFee"
                placeholder={t('send.fee.customPlaceholder')}
                error={error}
                transparent
                onInput={handleCustomFee}
                ref={inputRef}
                value={customFee}
              >
                <span className={style.customFeeUnit}>
                  { customFeeUnit(coinCode) }
                </span>
              </Input>
            </div>
          </div>
        )}
        { feeTarget && (
          <div>
            {(showCalculatingFeeLabel || proposeFeeText ? (
              <p className={style.feeProposed}>
                {t('send.fee.label')}:
                {' '}
                {showCalculatingFeeLabel ? t('send.feeTarget.placeholder') : proposeFeeText}
              </p>
            ) : null)}
            { !isCustom ? (
              <p className={style.feeDescription}>
                {t('send.feeTarget.estimate')}
                {' '}
                {t(`send.feeTarget.description.${feeTarget}`, {
                  context: getCoinCode(coinCode) || '',
                })}
              </p>
            ) : null }
          </div>
        )}
      </div>
    ) : (
      <Input
        disabled
        label={t('send.fee.label')}
        id="proposedFee"
        placeholder={t('send.fee.placeholder')}
        error={error}
        transparent
        value={proposeFeeText}
      />
    )
  );
};

