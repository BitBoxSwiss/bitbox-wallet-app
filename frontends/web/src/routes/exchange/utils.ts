/**
 * Copyright 2022-2024 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { IAccount } from '@/api/account';
import { getExchangeSupported, TExchangeName } from '@/api/exchanges';

/**
 * Gets formatted name for exchange.
 */
export const getExchangeFormattedName = (
  name: TExchangeName,
) => {
  switch (name) {
  case 'moonpay':
    return 'MoonPay';
  case 'pocket':
    return 'Pocket';
  case 'btcdirect':
    return 'BTC Direct';
  case 'btcdirect-otc':
    return 'BTC Direct\'s Private Trading Desk';
  case 'bitrefill':
    return 'Bitrefill';
  }
};

/**
 * Filters a given accounts list, keeping only the accounts supported by at least one exchange.
 */
export const getExchangeSupportedAccounts = async (accounts: IAccount[]): Promise<IAccount[]> => {
  const accountsWithFalsyValue = await Promise.all(
    accounts.map(async (account) => {
      const supported = await getExchangeSupported(account.code)();
      return supported.exchanges.length ? account : false;
    })
  );
  return accountsWithFalsyValue.filter(result => result) as IAccount[];
};
