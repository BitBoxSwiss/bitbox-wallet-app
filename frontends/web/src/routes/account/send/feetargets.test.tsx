/**
 * Copyright 2020 Shift Crypto AG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it, Mock, vi } from 'vitest';

vi.mock('../../../utils/request', () => ({
  apiGet: vi.fn().mockResolvedValue(''),
}));
vi.mock('../../../i18n/i18n');

import { render, waitFor } from '@testing-library/react';
import { FeeTargets } from './feetargets';
import { apiGet } from '../../../utils/request';
import * as apiHooks from '../../../hooks/api';

const useLoadSpy = vi.spyOn(apiHooks, 'useLoad');

describe('routes/account/send/feetargets', () => {

  it('should match the snapshot', async () => {
    useLoadSpy.mockReturnValue({ frontend: { expertFee: false } });
    (apiGet as Mock).mockResolvedValue({
      defaultFeeTarget: 'economy',
      feeTargets: [
        { code: 'low' },
        { code: 'economy' },
      ],
    });

    const { container } = render(
      <FeeTargets
        accountCode="btc"
        coinCode="btc"
        disabled={false}
        fiatUnit="USD"
        proposedFee={{
          amount: '1',
          unit: 'ETH',
          conversions: {
            AUD: '0.02',
            BRL: '12900',
            CAD: '0.02',
            CHF: '0.01',
            CNY: '0.08',
            CZK: '0.06',
            EUR: '0.02',
            GBP: '0.02',
            HKD: '19880',
            NOK: '0.02',
            ILS: '0.02',
            JPY: '1.30',
            KRW: '14.43',
            PLN: '2',
            RUB: '0.88',
            SEK: '0.1',
            SGD: '32233',
            USD: '0.02',
            BTC: '0.02',
          },
        }}
        customFee=""
        onCustomFee={vi.fn()}
        onFeeTargetChange={vi.fn()} />,
    );
    waitFor(() => expect(container).toMatchSnapshot());
  });

  it('should match the snapshot with empty feetargets', async () => {
    useLoadSpy.mockReturnValue({ frontend: { expertFee: false } });
    (apiGet as Mock).mockResolvedValue({
      defaultFeeTarget: '',
      feeTargets: [],
    });

    const { container } = render(
      <FeeTargets
        accountCode="eth"
        coinCode="eth"
        disabled={false}
        fiatUnit="EUR"
        customFee=""
        onCustomFee={vi.fn()}
        onFeeTargetChange={vi.fn()} />,
    );
    await waitFor(() => expect(container).toMatchSnapshot());
  });

  it('should call onFeeTargetChange with default', () => new Promise<void>(async done => {
    useLoadSpy.mockReturnValue({ frontend: { expertFee: false } });
    const apiGetMock = (apiGet as Mock).mockResolvedValue({
      defaultFeeTarget: 'normal',
      feeTargets: [
        { code: 'low' },
        { code: 'economy' },
      ],
    });

    const onFeeTargetChangeCB = (code: string) => {
      expect(code).toBe('normal');
      done();
    };

    render(
      <FeeTargets
        accountCode="btc"
        coinCode="btc"
        disabled={false}
        fiatUnit="USD"
        customFee=""
        onCustomFee={vi.fn()}
        onFeeTargetChange={onFeeTargetChangeCB} />,
    );
    await waitFor(() => expect(apiGetMock).toHaveBeenCalled());
  }));
});
