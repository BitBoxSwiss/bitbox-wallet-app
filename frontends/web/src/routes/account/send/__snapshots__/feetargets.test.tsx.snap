// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`routes/account/send/feetargets > should match the snapshot 1`] = `
<div>
  <div
    class="_input_44a2ea _align-left_44a2ea  _isTransparent_44a2ea"
  >
    <div
      class="flex flex-row flex-between"
    >
      <label
        class=""
        for="feetarget"
      >
        send.priority
      </label>
    </div>
    <input
      autocomplete="off"
      autocorrect="off"
      disabled=""
      id="feetarget"
      placeholder="send.feeTarget.placeholder"
      spellcheck="false"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`routes/account/send/feetargets > should match the snapshot with empty feetargets 1`] = `
<div>
  <div
    class="_input_44a2ea _align-left_44a2ea  _isTransparent_44a2ea"
  >
    <div
      class="flex flex-row flex-between"
    >
      <label
        class=""
        for="feetarget"
      >
        send.priority
      </label>
    </div>
    <input
      autocomplete="off"
      autocorrect="off"
      disabled=""
      id="feetarget"
      placeholder="send.feeTarget.placeholder"
      spellcheck="false"
      type="text"
      value=""
    />
  </div>
</div>
`;
